cli/connhelper/connhelper.go

```diff
@@ -0,0 +1,239 @@
+// Package connhelper provides helpers for connecting to a remote daemon host with custom logic.
+package connhelper
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"net"
+	"net/url"
+	"os"
+	"os/exec"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/docker/cli/cli/connhelper/ssh"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+)
+
+// ConnectionHelper allows to connect to a remote host with custom stream provider binary.
+type ConnectionHelper struct {
+	Dialer func(ctx context.Context, network, addr string) (net.Conn, error)
+	Host   string // dummy URL used for HTTP requests. e.g. "http://docker"
+}
+
+// GetConnectionHelper returns Docker-specific connection helper for the given URL.
+// GetConnectionHelper returns nil without error when no helper is registered for the scheme.
+// URL is like "ssh://me@server01".
+func GetConnectionHelper(daemonURL string) (*ConnectionHelper, error) {
+	u, err := url.Parse(daemonURL)
```

Comment:
You can actually get a unparsedHost here which is not parseable - has no schema.
Like `127.0.0.1:2364`

Reply:
`GetConnectionHelper returns nil without error when no helper is registered for the scheme.`

Reply:
I think, this is actually unrelated to what I'm talking about. Just a line below you check for `err` from `url.Parse()` and since there is a possibility the passed `daemonURL` is in `127.0.0.1:2345` format `url.Parse()` will rightfully return an error per its specification - no protocol scheme. I am not sure where should the scheme be added - here or higher in the call stack but this line in particular introduced a regression in the newest stable release of docker engine - 18.09.0

Reply:
@zmilonas looks like you mean the issue that was addressed by https://github.com/docker/cli/issues/1443 ?

Reply:
Well it might have been addressed but is not fixed 😅 I will file a new issue shortly then.

Reply:
It's fixed on master, which is what this repository is tracking.

A backport PR is open to include it in the 18.09 codebase

Reply:
Yes, that's what the backport is for. In the meantime you could either modify your configuration (use `tcp://<IP|host>:<port>`), or download the static binaries for Docker 18.06 (or "nightly", which has the latest changes from "master") https://download.docker.com/linux/static/